[staging] {cc,binutils}-wrapper: fixes for PIE hardening

r-burns opened the pull request:
Fixes a number of packages which were broken on musl (where PIE hardening is currently enabled). For example:
* pkgsMusl.python3
* pkgsMusl.bulletml
* pkgsMusl.proot
* pkgsMusl.libfsm
* pkgsMusl.libiscsi
* pkgsMusl.nsjail
* pkgsMusl.pv

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Fixes https://github.com/NixOS/nixpkgs/issues/124476

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

r-burns committed:
cc-wrapper: ensure PIE flags precede PIC flags

fixes:
pkgsMusl.bulletml
pkgsMusl.proot
pkgsMusl.python3

Debian explains this issue well in the dpkg-buildflags manpage:

-fPIE
    Can be linked into any program, but not a shared library (recommended).
-fPIC
    Can be linked into any program and shared library.

On projects that build both programs and shared libraries you might need to
make sure that when building the shared libraries -fPIC is always passed last
(so that it overrides any previous -PIE) to compilation flags such as CFLAGS.

(from https://manpages.debian.org/bullseye/dpkg-dev/dpkg-buildflags.1.en.html#hardening)

r-burns committed:
Revert "pkgsMusl.libiscsi: fix build"

This is no longer needed with the previous PIE hardening fixes.

This reverts commit 78d20f22007a9e615c8bf1a7bf12afea518e1d39.

r-burns commented:
cc @TredwellGit since it looks like you're running into some of these issues in https://github.com/NixOS/nixpkgs/pull/104091

r-burns committed:
Revert "glib: fix musl build"

This is no longer needed with the previous PIE hardening fixes.

This reverts commit 74e0aaaa30c1751450cddf055a5eb4ca64d7e33b.

r-burns committed:
Revert "pkgsMusl.python*: disable LTO"

This is no longer needed with the previous PIE hardening fixes.

This reverts commit 2784f1bd6908b37ed0d686778f347e63692c0f84.

veprbl commented on pkgs/build-support/cc-wrapper/add-hardening.sh:
```diff
@@ -34,6 +34,16 @@ if (( "${NIX_DEBUG:-0}" >= 1 )); then
+# Ensure PIE flags occur before PIC, if applicable
+if [ "${hardeningEnableMap[pie]-}" = 1 ]; then
+  if (( "${NIX_DEBUG:-0}" >= 1 )); then echo HARDENING: enabling CFlags -fPIE >&2; fi
+  hardeningCFlags+=('-fPIE')
```
Can we instead do
`hardeningCFlags =( '-fPIE' "${hardeningCFlags[@]}" )`

r-burns replied:
Great idea, that is SO much cleaner than what I have. Gonna run the rebuild gauntlet on this and get back to you.

r-burns replied:
Works great!

r-burns committed:
{cc,binutils}-wrapper: match leading/trailing arguments

fixes e.g.:
pkgsMusl.libfsm
pkgsMusl.libiscsi
pkgsMusl.nsjail
pkgsMusl.pv

match strings have whitespace on either side, which wasn't
matching leading/trailing arguments previously
